files/archive.py

```diff
@@ -0,0 +1,385 @@
+    required: false
+    default: null
+  remove:
+    description:
+      - Remove any added source files and trees after adding to archive.
+    type: bool
+    required: false
+    default: false
+
+author: "Ben Doherty (@bendoh)"
+notes:
+    - requires tarfile, zipfile, gzip, and bzip2 packages on target host
+    - can produce I(gzip), I(bzip2) and I(zip) compressed files or archives
+'''
+
+EXAMPLES = '''
+# Compress directory /path/to/foo/ into /path/to/foo.tgz
+- archive: path=/path/to/foo dest=/path/to/foo.tgz
+
+# Compress regular file /path/to/foo into /path/to/foo.gz and remove it
+- archive: path=/path/to/foo remove=True
+
+# Create a zip archive of /path/to/foo
+- archive: path=/path/to/foo compression=zip
+
+# Create a bz2 archive of multiple files, rooted at /path
+- archive:
+    path:
+        - /path/to/foo
+        - /path/wong/foo
+    dest: /path/file.tar.bz2
+    compression: bz2
+'''
+
+RETURN = '''
+state:
+    description:
+        The current state of the archived file.
+        If 'absent', then no source files were found and the archive does not exist.
+        If 'compress', then the file source file is in the compressed state.
+        If 'archive', then the source file or paths are currently archived.
+        If 'incomplete', then an archive was created, but not all source paths were found.
+    type: string
+    returned: always
+missing:
+    description: Any files that were missing from the source.
+    type: list
+    returned: success
+archived:
+    description: Any files that were compressed or added to the archive.
+    type: list
+    returned: success
+arcroot:
+    description: The archive root.
+    type: string
+expanded_paths:
+    description: The list of matching paths from paths argument.
+    type: list
+'''
+
+import stat
+import os
+import errno
+import glob
+import shutil
+import gzip
+import bz2
+import filecmp
+import zipfile
+import tarfile
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            path = dict(type='list', required=True),
+            compression = dict(choices=['gz', 'bz2', 'zip'], default='gz', required=False),
+            dest = dict(required=False),
+            remove = dict(required=False, default=False, type='bool'),
+        ),
+        add_file_common_args=True,
+        supports_check_mode=True,
+    )
+
+    params = module.params
+    paths = params['path']
+    dest = params['dest']
+    remove = params['remove']
+    expanded_paths = []
+    compression = params['compression']
+    globby = False
+    changed = False
+    state = 'absent'
+
+    # Simple or archive file compression (inapplicable with 'zip' since it's always an archive)
+    archive = False
+    successes = []
+
+    for i, path in enumerate(paths):
+        path = os.path.expanduser(path)
+
+        # Detect glob-like characters
+        if any((c in set('*?')) for c in path):
```

Comment:
This is probably quicker written as:
``` python
if ('*' in path or '?' in path):
```

The code you have is looping over every character in the path, creating a set each time, and then doing the containment test.  Simply checking for containment twice is probably faster.